### PR TITLE
Optimize io.cache ##io

### PR DESCRIPTION
**Pull request by GustavoLCR**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

* Use RPVector instead of RList
* Add cache skyline for faster reads
* Use PJ in `wcj`

Relocs killed r2's performance

#### Before

`r2 -e bin.dbginfo=0 -e io.cache=1 -qq libr/asm/libr_asm.so  93.41s user 0.19s system 100% cpu 1:33.59 total`

#### After

`r2 -ebin.dbginfo=0 -eio.cache=1 -qq libr/asm/libr_asm.so  0.97s user 0.16s system 100% cpu 1.120 total`

#### Without `-e io.cache=1`

`r2 -e bin.dbginfo=0 -qq libr/asm/libr_asm.so  0.77s user 0.14s system 99% cpu 0.914 total`

**Test plan**

Tests pass

### CODE REVIEW

File: libr/core/disasm.c [hunk 1]
```diff
@@ -6729,17 +6732,23 @@ R_API int r_core_disasm_pde(RCore *core, int nb_opcodes, int mode) {
...
 		}
-		core->io->cache->free = ocache->free;
+		core->io->cache.v.free = ocache.v.free;
 	}
 	r_io_cache_fini (core->io);
-	core->io->cache = ocache;
+	r_io_cache_init (core->io);
+	void **it;
+	r_pvector_foreach (&ocache, it) {
+		RIOCache *c = *it;
```
Comment (wargio) [inline libr/core/disasm.c:6748]: please add a cast here.

File: libr/io/io_cache.c [hunk 2]
```diff
@@ -21,37 +21,104 @@ static void cache_item_free(RIOCache *cache) {
...
-		if (r_itv_contain (c->itv, addr)) {
-			return true;
+#define CMP_END_GTE_PART(addr, part) \
+	(((addr) < (r_itv_end (((RIOCacheSkyline *)(part))->itv)) || !r_itv_end (((RIOCacheSkyline *)(part))->itv)) ? -1 : 1)
+
+#define CMP_BEGIN_GTE_PART(addr, part) \
+	(((addr) > (r_itv_begin (((RIOCacheSkyline *)(part))->itv))) - ((addr) < (r_itv_begin (((RIOCacheSkyline *)(part))->itv))))
+
+// This is copy-pasta from io_map.c
+static bool add_cache_to_skyline(RIO *io, RIOCache *cache) {
```
Comment (trufae) [inline libr/io/io_cache.c:31]: Can we have this skyline thing generic to avoid copypasta?

**Review by trufae — CHANGES_REQUESTED** [review]
Finally o(1) iocache 🥳

File: libr/io/io_cache.c (hunk 2 above)
Comment (GustavoLCR) [reply libr/io/io_cache.c]: @trufae created RSkyline

**Review by trufae — APPROVED** [review]
Can you add a test for io.cache, even if this pr is not adding a new feature, just improving the performance, i think its worth adding a test for variable trying to read and write in different addresses (allocated and non allocated). Apart from that the PR LGTM!

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@GustavoLCR there are some ASAN problems now:
```
[XX] /home/runner/work/radare2/radare2/test/bins/fuzzed <fuzz> /home/runner/work/radare2/radare2/test/bins/fuzzed/poc-radare2-invalid-memory-access-elf-758
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc aaa /home/runner/work/radare2/radare2/test/bins/fuzzed/poc-radare2-invalid-memory-access-elf-758
-- stdout

-- stderr
=================================================================
==27533==ERROR: AddressSanitizer: heap-use-after-free on address 0x61300000bca0 at pc 0x7fb8d30b46d0 bp 0x7ffeada474f0 sp 0x7ffeada474e0
READ of size 16 at 0x61300000bca0 thread T0
    #0 0x7fb8d30b46cf in r_skyline_add ../libr/util/skyline.c:36
    #1 0x7fb8d0fb23a1 in io_map_new ../libr/io/io_map.c:45
    #2 0x7fb8d0fb31e2 in io_map_add ../libr/io/io_map.c:134
    #3 0x7fb8d0fb3279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #4 0x7fb8ca6950ef in add_section ../libr/core/cbin.c:2679
    #5 0x7fb8ca69afd7 in bin_sections ../libr/core/cbin.c:3075
    #6 0x7fb8ca6a7e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #7 0x7fb8ca672dc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #8 0x7fb8caac7357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #9 0x7fb8caac9d01 in r_core_bin_load ../libr/core/cfile.c:651
    #10 0x7fb8d297a8aa in r_main_radare2 ../libr/main/radare2.c:1106
    #11 0x55d40bed503b in main ../binr/radare2/radare2.c:96
    #12 0x7fb8d1d110b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #13 0x55d40bed4acd in _start (/home/runner/bin/radare2+0x93acd)

0x61300000bca0 is located 288 bytes inside of 384-byte region [0x61300000bb80,0x61300000bd00)
freed by thread T0 here:
    #0 0x7fb8d37efffe in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dffe)
    #1 0x7fb8d30b017d in r_vector_insert ../libr/util/vector.c:139
    #2 0x7fb8d30b460f in r_skyline_add ../libr/util/skyline.c:31
    #3 0x7fb8d0fb23a1 in io_map_new ../libr/io/io_map.c:45
    #4 0x7fb8d0fb31e2 in io_map_add ../libr/io/io_map.c:134
    #5 0x7fb8d0fb3279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #6 0x7fb8ca6950ef in add_section ../libr/core/cbin.c:2679
    #7 0x7fb8ca69afd7 in bin_sections ../libr/core/cbin.c:3075
    #8 0x7fb8ca6a7e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #9 0x7fb8ca672dc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #10 0x7fb8caac7357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #11 0x7fb8caac9d01 in r_core_bin_load ../libr/core/cfile.c:651
    #12 0x7fb8d297a8aa in r_main_radare2 ../libr/main/radare2.c:1106
    #13 0x55d40bed503b in main ../binr/radare2/radare2.c:96
    #14 0x7fb8d1d110b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

previously allocated by thread T0 here:
    #0 0x7fb8d37efffe in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dffe)
    #1 0x7fb8d30b017d in r_vector_insert ../libr/util/vector.c:139
    #2 0x7fb8d30b4b5a in r_skyline_add ../libr/util/skyline.c:53
    #3 0x7fb8d0fb23a1 in io_map_new ../libr/io/io_map.c:45
    #4 0x7fb8d0fb31e2 in io_map_add ../libr/io/io_map.c:134
    #5 0x7fb8d0fb3279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #6 0x7fb8ca6950ef in add_section ../libr/core/cbin.c:2679
    #7 0x7fb8ca69afd7 in bin_sections ../libr/core/cbin.c:3075
    #8 0x7fb8ca6a7e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #9 0x7fb8ca672dc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #10 0x7fb8caac7357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #11 0x7fb8caac9d01 in r_core_bin_load ../libr/core/cfile.c:651
    #12 0x7fb8d297a8aa in r_main_radare2 ../libr/main/radare2.c:1106
    #13 0x55d40bed503b in main ../binr/radare2/radare2.c:96
    #14 0x7fb8d1d110b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: heap-use-after-free ../libr/util/skyline.c:36 in r_skyline_add
Shadow bytes around the buggy address:
  0x0c267fff9740: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c267fff9750: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c267fff9760: fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa fa
  0x0c267fff9770: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c267fff9780: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c267fff9790: fd fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd
  0x0c267fff97a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c267fff97b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c267fff97c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c267fff97d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c267fff97e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==27533==ABORTING

-- exit status: 1


[XX] /home/runner/work/radare2/radare2/test/bins/fuzzed <fuzz> /home/runner/work/radare2/radare2/test/bins/fuzzed/id146
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc aaa /home/runner/work/radare2/radare2/test/bins/fuzzed/id146
-- stdout

-- stderr
=================================================================
==27760==ERROR: AddressSanitizer: heap-use-after-free on address 0x608000025c38 at pc 0x7faa405516d0 bp 0x7ffe0b42c6c0 sp 0x7ffe0b42c6b0
READ of size 16 at 0x608000025c38 thread T0
    #0 0x7faa405516cf in r_skyline_add ../libr/util/skyline.c:36
    #1 0x7faa3e44f3a1 in io_map_new ../libr/io/io_map.c:45
    #2 0x7faa3e4501e2 in io_map_add ../libr/io/io_map.c:134
    #3 0x7faa3e450279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #4 0x7faa37b320ef in add_section ../libr/core/cbin.c:2679
    #5 0x7faa37b37fd7 in bin_sections ../libr/core/cbin.c:3075
    #6 0x7faa37b44e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #7 0x7faa37b0fdc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #8 0x7faa37f64357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #9 0x7faa37f66d01 in r_core_bin_load ../libr/core/cfile.c:651
    #10 0x7faa3fe178aa in r_main_radare2 ../libr/main/radare2.c:1106
    #11 0x55694958903b in main ../binr/radare2/radare2.c:96
    #12 0x7faa3f1ae0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #13 0x556949588acd in _start (/home/runner/bin/radare2+0x93acd)

0x608000025c38 is located 24 bytes inside of 96-byte region [0x608000025c20,0x608000025c80)
freed by thread T0 here:
    #0 0x7faa40c8cffe in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dffe)
    #1 0x7faa4054d17d in r_vector_insert ../libr/util/vector.c:139
    #2 0x7faa4055160f in r_skyline_add ../libr/util/skyline.c:31
    #3 0x7faa3e44f3a1 in io_map_new ../libr/io/io_map.c:45
    #4 0x7faa3e4501e2 in io_map_add ../libr/io/io_map.c:134
    #5 0x7faa3e450279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #6 0x7faa37b320ef in add_section ../libr/core/cbin.c:2679
    #7 0x7faa37b37fd7 in bin_sections ../libr/core/cbin.c:3075
    #8 0x7faa37b44e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #9 0x7faa37b0fdc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #10 0x7faa37f64357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #11 0x7faa37f66d01 in r_core_bin_load ../libr/core/cfile.c:651
    #12 0x7faa3fe178aa in r_main_radare2 ../libr/main/radare2.c:1106
    #13 0x55694958903b in main ../binr/radare2/radare2.c:96
    #14 0x7faa3f1ae0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

previously allocated by thread T0 here:
    #0 0x7faa40c8cffe in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dffe)
    #1 0x7faa4054e72e in r_vector_push ../libr/util/vector.c:185
    #2 0x7faa40551b75 in r_skyline_add ../libr/util/skyline.c:55
    #3 0x7faa3e44f3a1 in io_map_new ../libr/io/io_map.c:45
    #4 0x7faa3e4501e2 in io_map_add ../libr/io/io_map.c:134
    #5 0x7faa3e450279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #6 0x7faa37b320ef in add_section ../libr/core/cbin.c:2679
    #7 0x7faa37b37fd7 in bin_sections ../libr/core/cbin.c:3075
    #8 0x7faa37b44e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #9 0x7faa37b0fdc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #10 0x7faa37f64357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #11 0x7faa37f66d01 in r_core_bin_load ../libr/core/cfile.c:651
    #12 0x7faa3fe178aa in r_main_radare2 ../libr/main/radare2.c:1106
    #13 0x55694958903b in main ../binr/radare2/radare2.c:96
    #14 0x7faa3f1ae0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: heap-use-after-free ../libr/util/skyline.c:36 in r_skyline_add
Shadow bytes around the buggy address:
  0x0c107fffcb30: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fffcb40: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fffcb50: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 00 fa
  0x0c107fffcb60: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fffcb70: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
=>0x0c107fffcb80: fa fa fa fa fd fd fd[fd]fd fd fd fd fd fd fd fd
  0x0c107fffcb90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c107fffcba0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c107fffcbb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c107fffcbc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c107fffcbd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==27760==ABORTING

-- exit status: 1


[XX] /home/runner/work/radare2/radare2/test/bins/fuzzed <fuzz> /home/runner/work/radare2/radare2/test/bins/fuzzed/clusterfuzz-testcase-minimized-ia_fuzz-5754875154792448.dms
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc aaa /home/runner/work/radare2/radare2/test/bins/fuzzed/clusterfuzz-testcase-minimized-ia_fuzz-5754875154792448.dms
-- stdout

-- stderr
../libr/bin/format/pe/pe.c:3930:27: runtime error: signed integer overflow: 184549632 * 18 cannot be represented in type 'int'
=================================================================
==28132==ERROR: AddressSanitizer: heap-use-after-free on address 0x610000000658 at pc 0x7f9f9a1e36d0 bp 0x7ffd6ab163c0 sp 0x7ffd6ab163b0
READ of size 16 at 0x610000000658 thread T0
    #0 0x7f9f9a1e36cf in r_skyline_add ../libr/util/skyline.c:36
    #1 0x7f9f980e13a1 in io_map_new ../libr/io/io_map.c:45
    #2 0x7f9f980e13f2 in r_io_map_new ../libr/io/io_map.c:50
    #3 0x7f9f980d911a in r_io_open_at ../libr/io/io.c:176
    #4 0x7f9f917c3783 in io_create_mem_map ../libr/core/cbin.c:2629
    #5 0x7f9f917c3dce in add_section ../libr/core/cbin.c:2660
    #6 0x7f9f917c9fd7 in bin_sections ../libr/core/cbin.c:3075
    #7 0x7f9f917d6e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #8 0x7f9f917a1dc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #9 0x7f9f91bf6357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #10 0x7f9f91bf8d01 in r_core_bin_load ../libr/core/cfile.c:651
    #11 0x7f9f99aa98aa in r_main_radare2 ../libr/main/radare2.c:1106
    #12 0x560b0c44603b in main ../binr/radare2/radare2.c:96
    #13 0x7f9f98e400b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #14 0x560b0c445acd in _start (/home/runner/bin/radare2+0x93acd)

0x610000000658 is located 24 bytes inside of 192-byte region [0x610000000640,0x610000000700)
freed by thread T0 here:
    #0 0x7f9f9a91effe in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dffe)
    #1 0x7f9f9a1df17d in r_vector_insert ../libr/util/vector.c:139
    #2 0x7f9f9a1e360f in r_skyline_add ../libr/util/skyline.c:31
    #3 0x7f9f980e13a1 in io_map_new ../libr/io/io_map.c:45
    #4 0x7f9f980e13f2 in r_io_map_new ../libr/io/io_map.c:50
    #5 0x7f9f980d911a in r_io_open_at ../libr/io/io.c:176
    #6 0x7f9f917c3783 in io_create_mem_map ../libr/core/cbin.c:2629
    #7 0x7f9f917c3dce in add_section ../libr/core/cbin.c:2660
    #8 0x7f9f917c9fd7 in bin_sections ../libr/core/cbin.c:3075
    #9 0x7f9f917d6e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #10 0x7f9f917a1dc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #11 0x7f9f91bf6357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #12 0x7f9f91bf8d01 in r_core_bin_load ../libr/core/cfile.c:651
    #13 0x7f9f99aa98aa in r_main_radare2 ../libr/main/radare2.c:1106
    #14 0x560b0c44603b in main ../binr/radare2/radare2.c:96
    #15 0x7f9f98e400b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

previously allocated by thread T0 here:
    #0 0x7f9f9a91effe in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dffe)
    #1 0x7f9f9a1df17d in r_vector_insert ../libr/util/vector.c:139
    #2 0x7f9f9a1e3b5a in r_skyline_add ../libr/util/skyline.c:53
    #3 0x7f9f980e13a1 in io_map_new ../libr/io/io_map.c:45
    #4 0x7f9f980e21e2 in io_map_add ../libr/io/io_map.c:134
    #5 0x7f9f980e2279 in r_io_map_add_batch ../libr/io/io_map.c:145
    #6 0x7f9f917c40ef in add_section ../libr/core/cbin.c:2679
    #7 0x7f9f917c9fd7 in bin_sections ../libr/core/cbin.c:3075
    #8 0x7f9f917d6e96 in r_core_bin_info ../libr/core/cbin.c:4209
    #9 0x7f9f917a1dc0 in r_core_bin_set_env ../libr/core/cbin.c:345
    #10 0x7f9f91bf6357 in r_core_file_do_load_for_io_plugin ../libr/core/cfile.c:441
    #11 0x7f9f91bf8d01 in r_core_bin_load ../libr/core/cfile.c:651
    #12 0x7f9f99aa98aa in r_main_radare2 ../libr/main/radare2.c:1106
    #13 0x560b0c44603b in main ../binr/radare2/radare2.c:96
    #14 0x7f9f98e400b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: heap-use-after-free ../libr/util/skyline.c:36 in r_skyline_add
Shadow bytes around the buggy address:
  0x0c207fff8070: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa
  0x0c207fff8080: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c207fff8090: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c207fff80a0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c207fff80b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c207fff80c0: fa fa fa fa fa fa fa fa fd fd fd[fd]fd fd fd fd
  0x0c207fff80d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c207fff80e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c207fff80f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c207fff8100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c207fff8110: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==28132==ABORTING

-- exit status: 1
```

**Comment by XVilka** [comment]
Regarding skyline complexity, see the paper [Worst-case I/O-efficient Skyline Algorithms](https://github.com/radareorg/radare2/files/5581462/tods12-sky.pdf).

**Comment by trufae** [comment]
Unmerge if its introducing segfaults

File: libr/core/disasm.c [hunk 3]
```diff
@@ -6729,17 +6732,23 @@ R_API int r_core_disasm_pde(RCore *core, int nb_opcodes, int mode) {
...
-			(void)r_list_pop_head (core->io->cache);
+			(void)r_pvector_pop_front (&core->io->cache);
 			len--;
 		}
-		core->io->cache->free = ocache->free;
+		core->io->cache.v.free = ocache.v.free;
 	}
 	r_io_cache_fini (core->io);
 	core->io->cache = ocache;
+	r_skyline_clear (&core->io->cache_skyline);
```
Comment (ret2libc) [inline libr/core/disasm.c:6746]: I wonder why there is this messing around with caches in `disasm.c`. It is not exactly ideal.

**Review by ret2libc — COMMENTED** [review]
Awesome stuff! It would have been super cool to also have small unit tests for the new r_skyline, but awesome thing nonetheless :) I guess opening kernels now will be much much faster!! Thanks

**Comment by XVilka** [comment]
I am not sure if it's related to this PR, but seems there is a new IO-related timeout:
```

[XX] TIMEOUT db/io/zip io-zip hello.apk
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc ieq apk://bins/dex/Hello.apk
-- stdout
@@ -1 +0,0 @@
-0x000001c0

-- exit status: -1
```

**Pull request merged**